### PR TITLE
Filename column + hive partitioning parsing

### PR DESCRIPTION
**Pull request by samansmink**
This PR introduces 2 new features:

Firstly, a filename option is added for the parquet reader, similarly to the option that already existed for the CSV reader. Setting this option will add a column containing the filename to a scan/copy of the parquet reader. This is useful when scanning multiple files, for example when using a glob.

Secondly, a simple hive partitioning parsing feature is added to the parquet and csv readers, similar to what [bigquery does](https://cloud.google.com/bigquery/docs/hive-partitioned-queries-gcs). We don't do type casting yet though, the partitions are just left as strings for now leaving any desired casting up to the user. 

For example with the files:
```
./part=a/date=2012-01-01/test.parquet
./part=b/date=2013-01-01/test.parquet
```
The query 
```sql
select part, date from parquet_scan('./*/*/test.parquet', HIVE_PARTITIONING=1);
```
will return
```sql
a	2012-01-01
b	2013-01-01
```
Note that both test.parquet files do not actually contain a part and date, these are generated by the parquet reader.

A nice next step would be to prune the list of files based on the parsed hive partitions or filename column when there's a filter on one of these columns. Currently filtering on partition/filename columns in the csv reader is very inefficient as the files will be scanned anyway. For Parquet, it is less bad since row-group skipping will mean that only the metadata is read unnecessarily. However for example when scanning many files over the network with a selective filter on a hive/filename column, this will still add up quite a bit.

### CODE REVIEW

File: src/include/duckdb/common/hive_partitioning.hpp [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace duckdb {
+// matches hive partitions in file name. For example:
+// 	- s3://bucket/var1=value1/bla/bla/var2=value2
+//  - http(s)://domain(:port)/lala/kasdl/var1=value1/?not-a-var=not-a-value
+//  - folder/folder/folder/../var1=value1/etc/.//var2=value2
+inline std::map<string, string> ParseHivePartitions(string filename) {
+	std::map<string, string> result;
+
+	string regex = "[\\/\\\\]([^\\/\\?\\\\]+)=([^\\/\\n\\?\\\\]+)";
```
Comment (Mytherin) [inline src/include/duckdb/common/hive_partitioning.hpp:21]: What a beauty :)

File: extension/parquet/parquet_reader.cpp [hunk 2]
```diff
@@ -343,16 +345,33 @@ unique_ptr<ColumnReader> ParquetReader::CreateReader(const duckdb_parquet::forma
...
+	if (parquet_options.filename) {
+		Value val = Value(file_name);
+		root_struct_reader.child_readers.push_back(make_unique<GeneratedConstantColumnReader>(
+		    *this, LogicalType::VARCHAR, SchemaElement(), next_file_idx, 0, 0, val));
+	}
+
+	if (parquet_options.hive_partitioning) {
+		auto res = ParseHivePartitions(file_name);
+
+		for (auto &partition : res) {
```
Comment (Mytherin) [inline extension/parquet/parquet_reader.cpp:368]: Can we add a test for a mismatch in the number of hive partitions, e.g.:

```sql
SELECT * FROM read_parquet([
   'data/normal_folder/test.parquet',
   'data/a=1/test.parquet'
], HIVE_PARTITIONING=1);
```

Or:

```sql
SELECT * FROM read_csv([
   'data/*/test.csv'
], HIVE_PARTITIONING=1);
```

(where we have two files `data/normal_folder/test.csv` and `data/a=1/test.csv`)

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks great, minor comments:

File: src/include/duckdb/common/hive_partitioning.hpp (hunk 1 above)
Comment (samansmink) [reply src/include/duckdb/common/hive_partitioning.hpp:21]: 😅

**Comment by eitsupi** [comment]
Seems related to #2046 and #2186.

**Pull request merged**